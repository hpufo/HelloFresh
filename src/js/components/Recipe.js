import React from 'react';
import '../../scss/Recipes.scss';

export default class Recipe extends React.Component{
  constructor(props){
    super(props);
    this.state = {
      stars: 0,
      favorite: false
    }
  }
  onRating(rating){
    this.setState({stars: rating});
  }
  onFavorite = () => {
    this.setState({
      favorite: !this.state.favorite
    });
  }
  rednerRating(){
    let stars = [];
    for(let i=0; i<4; i++){
      let fill = this.state.stars >= i+1 ? "gold": "#b0aba3";
      stars.push(
        <svg width="16px" height="16px" onClick={() => this.onRating(i+1)} key={i}>
          <g fill="none">
            <path fill={fill} d="M8 11.646L11.708 14l-.981-4.44L14 6.576l-4.315-.39L8 2 6.315 6.185 2 6.575l3.273 2.986L4.292 14z" />
          </g>
        </svg>
      );
    }
    return stars;
  }
  render(){
    const {name,imageLink,headline,prepTime,calories} = this.props.recipe;
    const favSrc = this.state.favorite ? "filledStar" : "blankStar";
    
    return (
      <div className="recipe">
        <div className="imageContainer">
          <img className="image" src={imageLink} alt={name} />
          <img className="favorite" src={`images/${favSrc}.png`} onClick={this.onFavorite} alt="favorte recipe" />
        </div>
        <div className="recipeInfo">
          <h4>{name}</h4>
          <p className="headline">{headline}</p>
          <div className="metaInfo">
            <div className="infoContainer">
              <div className="calories">{calories} kcal</div>
              <div className="prepTime">{prepTime.match(/[0-9]*/g).join('')} minutes</div>
            </div>
            <div className="rating">{this.rednerRating()}</div>
          </div>
        </div>
      </div>
    );
  }
}