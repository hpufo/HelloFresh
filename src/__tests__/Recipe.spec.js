import React from 'react';
import {shallow} from 'enzyme';
import Recipe from '../js/components/Recipe';
import { shallowToJson } from 'enzyme-to-json';

let props = {
  name: "Test Name",
  imageLink: "test.png",
  headline: "Test headline",
  prepTime: "PT35M",
  calories: 355
}
describe('Recipe', () => {
  it('renders the props', () => {
    const wrapper = shallow(<Recipe recipe={props} />);
    expect(wrapper.find('h4').text()).toBe("Test Name");
    expect(wrapper.find('.headline').text()).toBe("Test headline");
    expect(wrapper.find('.calories').text()).toBe("355 kcal");
    expect(wrapper.find('.prepTime').text()).toBe("35 minutes");
    expect(wrapper.find('.image').prop('src')).toBe("test.png");
  });
  it('tests the favorite feature', () => {
    const wrapper = shallow(<Recipe recipe={props} />);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
    wrapper.find('.favorite').simulate('click');
    expect(wrapper.find('.favorite').prop('src')).toBe("images/filledStar.png");
    wrapper.find('.favorite').simulate('click');
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
  
  it('tests the ratings', () => {
    const wrapper = shallow(<Recipe recipe={props} />);
    wrapper.setState({stars: 2});
    expect(wrapper.find('path[fill="gold"]').length).toBe(2);
  });
});