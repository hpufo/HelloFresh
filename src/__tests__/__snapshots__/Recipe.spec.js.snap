// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recipe tests the favorite feature 1`] = `
<div
  className="recipe"
>
  <div
    className="imageContainer"
  >
    <img
      alt="Test Name"
      className="image"
      src="test.png"
    />
    <img
      alt="favorte recipe"
      className="favorite"
      onClick={[Function]}
      src="images/blankStar.png"
    />
  </div>
  <div
    className="recipeInfo"
  >
    <h4>
      Test Name
    </h4>
    <p
      className="headline"
    >
      Test headline
    </p>
    <div
      className="metaInfo"
    >
      <div
        className="infoContainer"
      >
        <div
          className="calories"
        >
          355
           kcal
        </div>
        <div
          className="prepTime"
        >
          35
           minutes
        </div>
      </div>
      <div
        className="rating"
      >
        <svg
          height="16px"
          key="0"
          onClick={[Function]}
          width="16px"
        >
          <g
            fill="none"
          >
            <path
              d="M8 11.646L11.708 14l-.981-4.44L14 6.576l-4.315-.39L8 2 6.315 6.185 2 6.575l3.273 2.986L4.292 14z"
              fill="#b0aba3"
            />
          </g>
        </svg>
        <svg
          height="16px"
          key="1"
          onClick={[Function]}
          width="16px"
        >
          <g
            fill="none"
          >
            <path
              d="M8 11.646L11.708 14l-.981-4.44L14 6.576l-4.315-.39L8 2 6.315 6.185 2 6.575l3.273 2.986L4.292 14z"
              fill="#b0aba3"
            />
          </g>
        </svg>
        <svg
          height="16px"
          key="2"
          onClick={[Function]}
          width="16px"
        >
          <g
            fill="none"
          >
            <path
              d="M8 11.646L11.708 14l-.981-4.44L14 6.576l-4.315-.39L8 2 6.315 6.185 2 6.575l3.273 2.986L4.292 14z"
              fill="#b0aba3"
            />
          </g>
        </svg>
        <svg
          height="16px"
          key="3"
          onClick={[Function]}
          width="16px"
        >
          <g
            fill="none"
          >
            <path
              d="M8 11.646L11.708 14l-.981-4.44L14 6.576l-4.315-.39L8 2 6.315 6.185 2 6.575l3.273 2.986L4.292 14z"
              fill="#b0aba3"
            />
          </g>
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`Recipe tests the favorite feature 2`] = `
<div
  className="recipe"
>
  <div
    className="imageContainer"
  >
    <img
      alt="Test Name"
      className="image"
      src="test.png"
    />
    <img
      alt="favorte recipe"
      className="favorite"
      onClick={[Function]}
      src="images/blankStar.png"
    />
  </div>
  <div
    className="recipeInfo"
  >
    <h4>
      Test Name
    </h4>
    <p
      className="headline"
    >
      Test headline
    </p>
    <div
      className="metaInfo"
    >
      <div
        className="infoContainer"
      >
        <div
          className="calories"
        >
          355
           kcal
        </div>
        <div
          className="prepTime"
        >
          35
           minutes
        </div>
      </div>
      <div
        className="rating"
      >
        <svg
          height="16px"
          key="0"
          onClick={[Function]}
          width="16px"
        >
          <g
            fill="none"
          >
            <path
              d="M8 11.646L11.708 14l-.981-4.44L14 6.576l-4.315-.39L8 2 6.315 6.185 2 6.575l3.273 2.986L4.292 14z"
              fill="#b0aba3"
            />
          </g>
        </svg>
        <svg
          height="16px"
          key="1"
          onClick={[Function]}
          width="16px"
        >
          <g
            fill="none"
          >
            <path
              d="M8 11.646L11.708 14l-.981-4.44L14 6.576l-4.315-.39L8 2 6.315 6.185 2 6.575l3.273 2.986L4.292 14z"
              fill="#b0aba3"
            />
          </g>
        </svg>
        <svg
          height="16px"
          key="2"
          onClick={[Function]}
          width="16px"
        >
          <g
            fill="none"
          >
            <path
              d="M8 11.646L11.708 14l-.981-4.44L14 6.576l-4.315-.39L8 2 6.315 6.185 2 6.575l3.273 2.986L4.292 14z"
              fill="#b0aba3"
            />
          </g>
        </svg>
        <svg
          height="16px"
          key="3"
          onClick={[Function]}
          width="16px"
        >
          <g
            fill="none"
          >
            <path
              d="M8 11.646L11.708 14l-.981-4.44L14 6.576l-4.315-.39L8 2 6.315 6.185 2 6.575l3.273 2.986L4.292 14z"
              fill="#b0aba3"
            />
          </g>
        </svg>
      </div>
    </div>
  </div>
</div>
`;
